untrusted/dcap/src/lib.rs

```diff
@@ -0,0 +1,192 @@
+// Copyright (c) 2022 The MobileCoin Foundation
+//! Rust wrappers for DCAP (Data Center Attestation Primitives) quote
+//! generation
+
+use mc_sgx_dcap_sys::{
+    quote3_error_t, sgx_qe_cleanup_by_policy, sgx_qe_get_quote, sgx_qe_get_quote_size,
+    sgx_qe_get_target_info, sgx_ql_path_type_t, sgx_ql_set_path, sgx_report_t, sgx_target_info_t,
+};
+use mc_sgx_urts::Enclave;
+use std::ffi::CString;
+use std::mem;
+use std::mem::MaybeUninit;
+
+/// Errors generating quotes
+#[derive(Debug, PartialEq)]
+pub enum Error {
+    // An error provided from the SGX SDK
+    SgxStatus(quote3_error_t),
+}
+
+impl From<mc_sgx_urts::Error> for Error {
+    fn from(err: mc_sgx_urts::Error) -> Self {
+        match err {
+            mc_sgx_urts::Error::SgxStatus(x) => {
+                // TODO re-think, these codes don't transfer 1:1
+                //  However SGX keeps the code separate by providing a mask on
+                //  the upper bits of a 16 bit value per the below macro
+                //  `#define SGX_???_ERROR(x) (0x0000?000|(x))`  The `?` changes
+                //  based on where the error comes from
+                let error_code = x.0;
+                let quote3_error = quote3_error_t(error_code);
+                Error::SgxStatus(quote3_error)
+            }
+            mc_sgx_urts::Error::NoReportFunction => {
+                Error::SgxStatus(quote3_error_t::SGX_QL_UNABLE_TO_GENERATE_REPORT)
+            }
+        }
+    }
+}
+
+/// A quote from the SGX interface
+pub struct Quote {
+    // Only used in tests so far
+    #[allow(dead_code)]
+    quote: Vec<u8>,
+}
+
+impl Quote {
+    /// Returns a quote for the provided [Enclave].
+    ///
+    /// # Arguments
+    /// - `enclave` The enclave to generate the quote for.
+    pub fn new(enclave: &Enclave) -> Result<Self, Error> {
+        Self::load_in_process_enclaves()?;
+        // TODO need to have a common type instead of transmuting these
+        let target_info = Self::get_target_info()?;
+        let target_info = unsafe { mem::transmute(target_info) };
+        let report = enclave.create_report(Some(&target_info))?;
+        let report = unsafe { mem::transmute(report) };
+        let quote = Self::get_quote(report);
+        Self::cleanup_in_process_enclaves();
+        quote
+    }
+
+    fn load_in_process_enclaves() -> Result<(), Error> {
+        //TODO this should be guarded by a feature and this should only be done
+        //  once, maybe lazy_static
```

Comment:
https://docs.rs/once_cell seems like a better fit for that

Reply:
nice call out, I like how you can have local variables with it